src/Disks/IO/IOUringReader.cpp

```diff
@@ -0,0 +1,242 @@
+#include <Common/Stopwatch.h>
+#include <Common/setThreadName.h>
+#include <base/errnoToString.h>
+#include <future>
+
+namespace ProfileEvents
+{
+    extern const Event ReadBufferFromFileDescriptorRead;
+    extern const Event ReadBufferFromFileDescriptorReadFailed;
+    extern const Event ReadBufferFromFileDescriptorReadBytes;
+
+    extern const Event IOUringSQEQueueFullRetries;
+    extern const Event IOUringSQEsSubmitted;
+    extern const Event IOUringSQEsResubmits;
+    extern const Event IOUringShortReads;
+}
+
+namespace CurrentMetrics
+{
+    extern const Metric IOUringEnqueuedEvents;
+    extern const Metric Read;
+}
+
+namespace DB
+{
+
+namespace ErrorCodes
+{
+    extern const int LOGICAL_ERROR;
+    extern const int CANNOT_READ_FROM_FILE_DESCRIPTOR;
+    extern const int IO_URING_INIT_FAILED;
+    extern const int IO_URING_QUEUE_FULL;
+    extern const int IO_URING_WAIT_ERROR;
+}
+
+IOUringReader::IOUringReader(size_t queue_size_)
+{
+    struct io_uring_probe * probe = io_uring_get_probe();
+    if (!probe)
+    {
+        is_supported = false;
+        return;
+    }
+
+    is_supported = io_uring_opcode_supported(probe, IORING_OP_READ);
+    io_uring_free_probe(probe);
+
+    if (!is_supported)
+        return;
+
+    int ret = io_uring_queue_init(queue_size_, &ring, 0);
+    if (ret < 0)
+        throwFromErrno(fmt::format("Failed initializing io_uring of size {}", queue_size_), ErrorCodes::IO_URING_INIT_FAILED);
+
+    ring_completion_monitor = ThreadFromGlobalPool([this] { monitorRing(); });
+}
+
+std::future<IAsynchronousReader::Result> IOUringReader::submit(Request request)
+{
+    assert(request.size);
+
+    // take lock here because we're modifying an std::map and submitting to the ring,
+    // the monitor thread can also modify the map and re-submit events
+    std::unique_lock lock{mutex};
+
+    // use the requested read destination address as the request id, the assumption
+    // here is that we won't get asked to fill in the same address more than once in parallel
+    auto request_id = reinterpret_cast<UInt64>(request.buf);
+
+    const auto [kv, is_newly_inserted] = enqueued_requests.emplace(request_id, EnqueuedRequest{
+        .promise = std::promise<IAsynchronousReader::Result>{},
+        .request = request,
+        .bytes_read = 0
+    });
+
+    if (!is_newly_inserted)
+        return makeFailedResult(ErrorCodes::LOGICAL_ERROR, "Tried enqueuing read request for %lu that is already submitted", request_id);
+
+    EnqueuedRequest & enqueued = kv->second;
+    trySubmitRequest(request_id, enqueued, false);
+
+    return enqueued.promise.get_future();
+}
+
+bool IOUringReader::trySubmitRequest(UInt64 request_id, EnqueuedRequest & enqueued, bool resubmitting)
+{
+    auto request = enqueued.request;
+    int fd = assert_cast<const LocalFileDescriptor &>(*request.descriptor).fd;
+
+    struct io_uring_sqe * sqe = nullptr;
+    for (int i = 0; i < 1000; ++i) // try a few times in case the queue is full.
```

Comment:
Does it have some backoff?
Busy loop does not look reasonable...

Reply:
what about doing an atomic `wait` here and `notify_all` upon completions?

Reply:
turns out I misunderstood how this part works: sqe retrieval here can't fail in principle because `io_uring_submit` is called immediately afterwards and once it returns this sqe can be reused.